paddle/framework/framework.proto

```diff
@@ -91,52 +91,56 @@ message OpProto {
   required string comment = 5;
 }
 
-enum DataType {
-  BOOL = 0;
-  INT16 = 1;
-  INT32 = 2;
-  INT64 = 3;
-  FP16 = 4;
-  FP32 = 5;
-  FP64 = 6;
-}
-
-message TensorDesc {
-  required DataType data_type = 1;
-  repeated int64 dims = 2; // [UNK, 640, 480] is saved as [-1, 640, 480]
-}
-
-message LoDTensorDesc {
-  required TensorDesc tensor = 1;
-  optional int32 lod_level = 2 [ default = 0 ];
-}
-
-message LoDTensorArrayDesc {
-  required TensorDesc tensor = 1;
-  optional int32 lod_level = 2 [ default = 0 ];
-}
-
-message ReaderDesc { repeated LoDTensorDesc lod_tensor = 1; }
-
-message VarDesc {
-  enum VarType {
-    LOD_TENSOR = 1;
-    SELECTED_ROWS = 2;
+message VarType {
+  enum Type {
+    // POD types
+    BOOL = 0;
+    INT16 = 1;
+    INT32 = 2;
+    INT64 = 3;
+    FP16 = 4;
+    FP32 = 5;
+    FP64 = 6;
+
+    // other types, that might need additonal descriptions
+    LOD_TENSOR = 1; // LoDTensorType
+    SELECTED_ROWS = 2; // SelectedRowsType
     FEED_MINIBATCH = 3;
     FETCH_LIST = 4;
     STEP_SCOPES = 5;
     LOD_RANK_TABLE = 6;
-    LOD_TENSOR_ARRAY = 7;
-    PLACE_LIST = 8;
-    READER = 9;
+    // LOD_TENSOR_ARRAY = 7; // replaced by array-of-tensors. c.f. ArrayType
+    // PLACE_LIST = 8; // replaced by array-of-place. c.f. ArrayType
+    READER = 9; // ReaderType
+  }
+
+  required Type type = 1;
+  
+  message Tensor {
+    required Type data_type = 1;
```

Comment:
Is the data_type here for Tensor only POD types？

Reply:
As suggested by @wangkuiyi , one way could be to leave it as it is and check in the C++ code that `data_type` must be a POD. Another way could be to define this inside protobuf. We can split `Type` into 2 different enums :

```protobuf
enum PodType {
    // POD types
    BOOL = 0;
    INT16 = 1;
    INT32 = 2;
    INT64 = 3;
    FP16 = 4;
    FP32 = 5;
    FP64 = 6;
}

enum CompositeType {
    LOD_TENSOR = 1;    // LoDTensor
    SELECTED_ROWS = 2; // Tensor
    FEED_MINIBATCH = 3;
    FETCH_LIST = 4;
    STEP_SCOPES = 5;
    LOD_RANK_TABLE = 6;
    // LOD_TENSOR_ARRAY = 7; // replaced by array-of-tensors. c.f. ArrayType
    // PLACE_LIST = 8; // replaced by array-of-place. c.f. ArrayType
    READER = 7; // ReaderType
}
```

Once we have done this, we can define a field `type` in `VarType` as follows:

```protobuf
message VarType {
    oneof type {
        PodType ptype = 1;
        CompositeType ctype = 2;
    }
}
```protobuf

This way we can capture the requirement of our Tensor and LODTensor inside protobuf message. The tensor can be defined as follows:

```protobuf
message Tensor {
    required PodType data_type = 1;
    repeated int64 dims = 2; // [UNK, 640, 480] is saved as [-1, 640, 480]
  }
```